File: runme/mcgen/local/datan/datsrc/minpow.f
      SUBROUTINE MINPOW(X0,N,NRED,LIST,USERFN,FMIN,EPSILN,
     +NSTEP,ALLDIR)
      PARAMETER(MAXN=50)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION X0(N),XOLD(MAXN),XE(MAXN),DIR(MAXN),DIRN(MAXN)
      DIMENSION LIST(N),ALLDIR(N,NRED)
      EXTERNAL USERFN
      PARAMETER(MAXSTP=1000,EPSDEF=1.D-8,T=1.D-15,ONE=1.D0,
     +ZERO=0.D0)
      IF(NSTEP.LT.1) NSTEP=MAXSTP
      EPS=EPSILN
      IF(EPS.LE.0.) EPS=EPSDEF
C Initialize directions along coordinates
      F=USERFN(X0,N)
      IRED=0
      DO 10 I=1,N
        IF(LIST(I).EQ.1) THEN
          IRED=IRED+1
          CALL MTXZRV(DIR,N)
          DIR(I)=ONE
          CALL MTXPCL(ALLDIR,DIR,N,NRED,IRED)
        END IF
   10 CONTINUE
      DO 30 ISTEP=1,NSTEP
        LSTEP=ISTEP
        FOLD=F
        FL=FOLD
        CALL MTXCPV(X0,XOLD,N)
        DELTAF=0.
        IMAX=0
C Loop over complete direction set and find direction which
C yields largest function decrease. Its index is IMAX
        DO 20 I=1,NRED
          CALL MTXGCL(ALLDIR,DIR,N,NRED,I)
          NSTDIR=NSTEP
          CALL MINDIR(X0,DIR,N,USERFN,F,NSTDIR,ZERO)
          IF(NSTDIR.LT.0) GO TO 40
          D=ABS(F-FL)
          IF(D.GT.DELTAF) THEN
            DELTAF=D
            IMAX=I
          END IF
          FL=F
   20   CONTINUE
C Test for break-off criterion
        IF(ABS(FL-FOLD) .LT. EPS*ABS(FL)+T) GO TO 50
C Construct extrapolated point XE and direction DIRN from
C XOLD to X0
        CALL MTXSBV(X0,XOLD,DIRN,N)
        CALL MTXADV(DIRN,X0,XE,N)
        FE=USERFN(XE,N)
C Now there are 3 points (XOLD, X0, XE)
C and their function values (FOLD, F, FE)
        IF(FE.LT.F) THEN
          TEST=2.*(FOLD-2.*F+FE)*(FOLD-F-DELTAF)**2-
     +         (FOLD-FE)**2*DELTAF
          IF(TEST.LT.0.) THEN
            NST=NSTEP
C Find minimum along DIRN and replace X0 by position
C of minimum. Replace direction with index IMAX by DIRN
            CALL MINDIR(X0,DIRN,N,USERFN,F,NST,ZERO)
            IF(NST.LT.0) GO TO 40
            CALL MTXPCL(ALLDIR,DIRN,N,NRED,IMAX)
          END IF
        END IF
   30 CONTINUE
   40 NSTEP=-1
      GO TO 60
   50 NSTEP=LSTEP
      FMIN=FL
   60 CONTINUE
      END

File: runme/mcgen/local/datan/datsrc/minpow.c
#include <stdlib.h>
#include <limits.h>
#include "datsrc.h"

void LIBFUNC minpow_(double P_T *x0, integer const P_T *n,
             integer const P_T *nred, integer const P_T *list,
             double (CBFUNC P_T *userfn)(double const P_T *,
             integer const P_T *), double P_T *fmin,
             double const P_T *epsiln, integer P_T *nstep, double P_T *alldir)
{
    /* Table of constant values */

    double c_d0 = 0.;

    /* System generated locals */
    integer alldir_dim1, alldir_offset, i__1, i__2;
    double d__1, d__2;

    /* Local variables */
    integer i, istep, lstep, imax, ired, nst, nstdir;
    double eps, fe, fl, deltaf, fold, test, d, f;
    /*double dirn[50], xold[50], xe[50], dir[50];*/
    double *dir, *dirn, *xe, *xold;
    unsigned long amem;

    if ((amem = sizeof(double) * 5 * *n) > MAXALLOC) {
        MEMERROR("minpow_ (too large)");
        goto f_0;
    }
    dir = (double *)MEMALLOC(amem);
    if (!dir) {
        MEMERROR("minpow_");
        goto f_0;
    }
    dirn = dir + *n; xe = dirn + *n; xold = xe + *n;

    /* Parameter adjustments */
    --list;
    --x0;
    alldir_dim1 = *n;
    alldir_offset = alldir_dim1 + 1;
    alldir -= alldir_offset;

    /* Function Body */
    if (*nstep < 1) {
        *nstep = 1000;
    }
    eps = *epsiln;
    if (eps <= 0.f) {
        eps = 1e-8;
    }
/* Initialize directions along coordinates */
    f = (*userfn)(&x0[1], n);
    ired = 0;
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        if (list[i] == 1) {
            ++ired;
            mtxzrv_(dir, n);
            dir[i - 1] = 1.;
            mtxpcl_(&alldir[alldir_offset], dir, n, nred, &ired);
        }
    }
    i__1 = *nstep;
    for (istep = 1; istep <= i__1; ++istep) {
        lstep = istep;
        fold = f;
        fl = fold;
        mtxcpv_(&x0[1], xold, n);
        deltaf = 0.f;
        imax = 0;
/* Loop over complete direction set and find direction which */
/* yields largest function decrease. Its index is IMAX */
        i__2 = *nred;
        for (i = 1; i <= i__2; ++i) {
            mtxgcl_(&alldir[alldir_offset], dir, n, nred, &i);
            nstdir = *nstep;
            mindir_(&x0[1], dir, n, userfn, &f, &nstdir, &c_d0);
            if (nstdir < 0) {
                goto L40;
            }
            d = (d__1 = f - fl, abs(d__1));
            if (d > deltaf) {
                deltaf = d;
                imax = i;
            }
            fl = f;
        }
/* Test for break-off criterion */
        if ((d__1 = fl - fold, abs(d__1)) < eps * abs(fl) + 1e-15) {
            goto L50;
        }
/* Construct extrapolated point XE and direction DIRN from */
/* XOLD to X0 */
        mtxsbv_(&x0[1], xold, dirn, n);
        mtxadv_(dirn, &x0[1], xe, n);
        fe = (*userfn)(xe, n);
/* Now there are 3 points (XOLD, X0, XE) */
/* and their function values (FOLD, F, FE) */
        if (fe < f) {
/* Computing 2nd power */
            d__1 = fold - f - deltaf;
/* Computing 2nd power */
            d__2 = fold - fe;
            test = (fold - f * 2.f + fe) * 2.f * (d__1 * d__1) - d__2 * d__2 *
                     deltaf;
            if (test < 0.f) {
                nst = *nstep;
/* Find minimum along DIRN and replace X0 by position */
/* of minimum. Replace direction with index IMAX by DIRN */
                mindir_(&x0[1], dirn, n, userfn, &f, &nst, &c_d0);
                if (nst < 0) {
                    goto L40;
                }
                mtxpcl_(&alldir[alldir_offset], dirn, n, nred, &imax);
            }
        }
    }
L40:
    *nstep = -1;
    goto L60;
L50:
    *nstep = lstep;
    *fmin = fl;
L60:
    MEMFREE((void *)dir);
f_0:
    return;
} /* minpow_ */

